Fixed the Tiberian Sun weapon factory doors

Mailaender opened the pull request:
They will now open and show the insides properly with a new modular render overlay trait. `RenderBuildingWarFactory` was too hard-coded to the Red Alert and Tiberian Dawn sprites.

pchote commented on OpenRA.Mods.Common/Traits/Render/WithProductionDoor.cs:
```diff
@@ -0,0 +1,90 @@
+
+	class WithProductionDoor : INotifyBuildComplete, ITick, INotifyProduction, INotifySold, ISync, INotifyDamageStateChanged
+	{
+		WithProductionDoorInfo info;
```
`readonly`, please.

pchote commented on OpenRA.Mods.Common/Traits/Render/WithProductionDoor.cs:
```diff
@@ -0,0 +1,90 @@
+		WithProductionDoorInfo info;
+		RenderSprites renderSprites;
+		Animation door;
+		[Sync] bool isOpen;
```
This is a cosmetic trait, so these don't really need to be synced.

Mailaender replied:
Shall I also remove it from RenderBuildingWarFactory?

pchote replied:
Ok, I guess so.

pchote commented:
It might pay to talk to @Phrohdoh about what changes this will need for RA2 (the weapons factory has two doors with different animations), and if they aren't too difficult putting them in now.

Mailaender commented:
Updated.

Phrohdoh commented on OpenRA.Mods.Common/Traits/Render/WithProductionDoor.cs:
```diff
@@ -0,0 +1,91 @@
+{
+	class WithProductionDoorInfo : ITraitInfo, IRenderActorPreviewSpritesInfo, Requires<RenderSpritesInfo>, Requires<IBodyOrientationInfo>, Requires<BuildingInfo>
+	{
+		[Desc("Additional units relative to the center to align to the top of cell.")]
```
I can't understand this at all.

Mailaender replied:
Me neither. Removed it again. It shouldn't be configurable I guess.

abcdefg30 commented:
Code looks ok and works in TS. :+1: / :white_check_mark:

pchote commented on OpenRA.Mods.Common/Traits/Render/WithProductionDoor.cs:
```diff
@@ -0,0 +1,89 @@
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Play an animation when a unit exits or blocks the exit after production finished.")]
+	class WithProductionDoorInfo : ITraitInfo, IRenderActorPreviewSpritesInfo, Requires<RenderSpritesInfo>, Requires<IBodyOrientationInfo>, Requires<BuildingInfo>
```
The convention we have set is to call these traits `With*Overlay`.  You could rename this to `WithProductionDoorOverlay`, or merge its behaviour in to the existing `WithProductionOverlay` trait.

pchote commented on OpenRA.Mods.Common/Traits/Render/WithProductionDoor.cs:
```diff
@@ -0,0 +1,89 @@
+	[Desc("Play an animation when a unit exits or blocks the exit after production finished.")]
+	class WithProductionDoorInfo : ITraitInfo, IRenderActorPreviewSpritesInfo, Requires<RenderSpritesInfo>, Requires<IBodyOrientationInfo>, Requires<BuildingInfo>
+	{
+		public readonly string IdleSequence = "idle-door";
```
This should be just `Sequence`.

Mailaender committed:
readonly and don't sync cosmetic traits

Mailaender commented:
Rebased and renamed.

reaperrr commented:
Needs a rebase.

One question: Since (most likely) none of the currently shipping mods will use this and the RA2 mod will need ```OpenRA.Mods.TS``` anyway (because of stuff like voxels, and probably a couple of other things), wouldn't it make more sense to move this to the TS dll and namespace? On a sidenote, that should make this PR TS-only and no longer fall under the feature-freeze.

Mailaender committed:
fix the weapon factory door opening animation

Mailaender commented:
Rebased and moved to `OpenRA.Mods.TS`. I can't guarantee this will work with RA2, but I worked towards that it should also work with two door traits per building.

pchote commented:
@Mailaender suggested in IRC that this could be used to replace RenderBuildingWarFactory completely.  I agree, and would like to see this done before merging.

Mailaender committed:
deprecate RenderBuildingWarFactory

Mailaender commented:
Updated. This also fixes the Dune 2000 refinery being rendered twice. A regression from https://github.com/OpenRA/OpenRA/pull/7477. I always wondered why the shadow is so strong.

pchote commented:
Could we please get an upgrade rule that replaces `-RenderBuilding:` and `RenderBuildingWarFactory:` with `WithProductionDoorOverlay:`?  This will likely break custom maps otherwise.

Mailaender commented:
That is pretty tricky as that is not always the right update (see the fake buildings and refineries).

Mailaender committed:
add an upgrade rule

Mailaender committed:
remove trailing tabs

Mailaender commented:
Added an upgrade rule and ran it (did not forget any custom map in the repository).

pchote commented on OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs:
```diff
@@ -0,0 +1,90 @@
+			door.PlayThen(RenderSprites.NormalizeSequence(door, self.GetDamageState(), info.BuildSequence), () => { isOpen = true; openExit = exit; });
+		}
+
+		public void Selling(Actor self) { renderSprites.Remove("door_overlay_{0}".F(info.Sequence)); }
```
Convention is to make the overlay invisible (by setting a `buildComplete` bool to false) instead of explicitly removing the animation.

pchote commented on OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs:
```diff
@@ -0,0 +1,90 @@
+
+		public void BuildingComplete(Actor self)
+		{
+			door.Play(RenderSprites.NormalizeSequence(door, self.GetDamageState(), info.Sequence));
```
Convention is to do this in the constructor instead of here.

pchote commented on OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs:
```diff
@@ -0,0 +1,90 @@
+
+		public void UnitProduced(Actor self, Actor other, CPos exit)
+		{
+			door.PlayThen(RenderSprites.NormalizeSequence(door, self.GetDamageState(), info.BuildSequence), () => { isOpen = true; openExit = exit; });
```
This should only be played if `!isOpen`.

pchote replied:
It probably also wants to set `isOpen = true` at the start of the animation to avoid glitches.

pchote commented:
This adds some nice polish and code cleanups.  Still a couple more minor code issues to address before merging, though.

pchote commented on OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs:
```diff
@@ -0,0 +1,90 @@
+		{
+			if (isOpen && !self.World.ActorMap.GetUnitsAt(openExit).Any(a => a != self))
+			{
+				isOpen = false;
```
and then this should be set after the reverse animation plays.

Mailaender replied:
That won't work. It will then only close when a new actor is produced which is glitchy in a different kind of way and I can't get that fixed so I will keep the legacy behavior for now.

pchote replied:
Ok. You can solve the glitches in a different way by adding bools for `shouldClose = isOpen && self.World.ActorMap.GetUnitsAt(openExit).Any(a => a != self)` and `animating` = (set true at the start of an animation and false at the end).

The code here can then check `if (!animating && shouldClose)` and the check in `UnitProduced` becomes `if (!animating && !isOpen)`.

pchote replied:
Or perhaps simpler, maintain separate `isOpen` and `shouldBeOpen` bools, and then handle the transition in both directions from Tick.

Mailaender replied:
This is fiddly to get right. Tried all your suggestions, but it still doesn't work quite right especially with fast build cheats on. Let's do this in another pull request.

pchote committed:
Simplify and fix animation glitches.

pchote commented:
After some tweaking I came up with 04f8dda26a002b80dd7dca90e8a9e4468f931d8b as a simpler implementation that sidesteps the possibility of animation glitches.

Mailaender commented:
Integrated your commit here.

Mailaender commented:
I may be able to use your new animation method to complete https://github.com/OpenRA/OpenRA/pull/7998.

pchote commented:
:+1: from me, but my patch significantly changes this so i'll wait for somebody else to give a final review and merge.

penev92 commented on OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs:
```diff
@@ -0,0 +1,86 @@
+			anim.PlayFetchIndex(Sequence, () => 0);
+
+			var bi = init.Actor.Traits.Get<BuildingInfo>();
+			var offset = FootprintUtils.CenterOffset(init.World, bi).Y + 512; // Additional 512 units move from center -> top of cell
```
I really don't like the magic number here. The comment doesn't help me understand what is happening either.
Can we make this configurable (with a nice description maybe?), or at least add a better comment what is happening here and why we have a hadrcoded number.

pchote replied:
I'm not convinced we need to deal with that right now.This is legacy code that has been moved from the original trait.

My preference (for whenever this gets dealt with) would be to remove this completely and rely on a sequence-defined z-offset.

Mailaender replied:
We had it configurable, but it only confused @Phrohdoh and it also should not be changed. Did some unsuccessful experiments already weeks ago.

penev92 commented on OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs:
```diff
@@ -0,0 +1,86 @@
+
+			var buildingInfo = self.Info.Traits.Get<BuildingInfo>();
+
+			var offset = FootprintUtils.CenterOffset(self.World, buildingInfo).Y + 512;
```
Same thing is happening here, I see. Still no clue what it is.

Mailaender replied:
Okay, replaced the magic number with something else.

penev92 commented on mods/ts/sequences/structures.yaml:
```diff
@@ -329,7 +332,6 @@ naweap:
 	icon: nwepicon
 		Offset: 0, 0
 		UseTilesetCode: false
-# TODO: ntweap_1 & ntweap_b & ntweap_c are unused
```
I still don't see a `ntweap_c` anywhere.

penev92 commented:
Well apart from those leftover magic numbers the code looks simple enough (as I've been told, there are no spaghetti here).
Works ingame, so I'm on board.

:+1:  once that pesky `ntweap_c` is found.

Mailaender commented:
It doesn't exist... 
![image](https://cloud.githubusercontent.com/assets/756669/7333963/7e523316-eb82-11e4-95e8-1ae57ab48e62.png)

pchote commented on OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs:
```diff
@@ -0,0 +1,85 @@
+				() => desiredFrame - door.CurrentFrame);
+
+			var buildingInfo = self.Info.Traits.Get<BuildingInfo>();
+			var offset = FootprintUtils.CenterOffset(self.World, buildingInfo).Y + (new WRange(1).Range / 2); // move from center -> top of cell
```
This is even less readable than 512, IMO.  It also evaluates to 0, so is wrong.

Mailaender replied:
Removed it again.

penev92 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/b1ffeb667c28e2be24a3d70572deb8b6fb01d297...633b487f6eb93799a6e1454bbb2949fc6428fad6)

Mailaender commented:
[supplementing](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/633b487f6eb93799a6e1454bbb2949fc6428fad6...b026c715c8fcc8c012608930f17eae68b22aeb37)
